File: Terminal/Pdf.cs
using ExtensionMethods;
using Optional;
using PdfSharp.Drawing;
using PdfSharp.Pdf;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

using System.Text;
using ImageMagick;

public static class Pdf
{
    public static void GenerateNew(PathBundle pathBundle)
    {
        List<string> images = pathBundle.Paths;
        RequestInfo requestInfo = pathBundle.Metadata;
        string title = requestInfo.Title;
        int chapter = requestInfo.Id;

        using var document = new PdfDocument();
        document.AddPages(images.Count);

        images.ForEach((path, pageIndex) =>
        {
            using XImage image = LoadImageFromFile(path);
            PdfPage page = document.Pages[pageIndex];
            using XGraphics renderer = XGraphics.FromPdfPage(page);

            page.Width = image.PixelWidth;
            page.Height = image.PixelHeight;
            renderer.DrawImage(image, 0, 0, page.Width, page.Height);
        });

        document.Save($"{title}/{title} Chapter {chapter}.pdf");
        Directory.Delete($"{title}/{chapter}", true);
    }

    private static XImage LoadImageFromFile(string path)
    {
        try
        {
            return XImage.FromFile(path);
        }
        catch (InvalidOperationException)
        {
            using var image = new MagickImage(path);
            image.Write(path, MagickFormat.Png24);

            return XImage.FromFile(path);
        }
    }
}

File: src/backend/Pdf.fs
module Pdf

open CommonTypes
open PdfSharp.Drawing
open PdfSharp.Pdf
open System

let generateNew (chapter: Chapter) (name: string) : Unit =
    use document = new PdfDocument()
    document.AddPages (chapter.ImageInfo.Local.Length)

    chapter.ImageInfo.Local
    |> Seq.iteri (fun i path ->
        use image = XImage.FromFile(path)
        let page = document.Pages[i]

        page.Width <- XUnit.FromPoint(image.PixelWidth)
        page.Height <- XUnit.FromPoint(image.PixelHeight)

        use renderer = XGraphics.FromPdfPage(page)
        renderer.DrawImage(image, 0, 0, page.Width.Point, page.Height.Point))

    document.Info.Title <- name
    document.Info.Elements.Add("/Scanlators", PdfString("shit"))

    let file = String.Format("{0} Chapter {1}.pdf", name, chapter.Number.ToString "D4")
    document.Save (Utility.joinPathsUnix [| name; file |])
